### PR TITLE
fix: fail test when timeout(0) and promise never resolves

### LINKED ISSUE #2537: Inconsistent timeout policy
Timeout 0 and un-resolved async function lead nowhere.
No throw, no error, no next test running, nothing.

```
describe("should never end", function(){
  this.timeout(0);

  it("should never end", function(done){ });

  it("is here now", function(){
    console.log("Allgood");
  });
});
```

Current result 
```
R:\>node d:dvp\node_modules\mocha\bin\_mocha test.js
  should never end
R:\>
```
Expected result
```
R:\>node d:dvp\node_modules\mocha\bin\_mocha test.js
  should never end
  (???? something ?? waiting undefinitively ?) , but NOT nothing
R:\>
```

### PR DESCRIPTION
**Pull request by mostafaNazari702**
## PR Checklist

- [x] Addresses an existing open issue: fixes #2537
- [x] That issue was marked as [`status: accepting prs`](https://github.com/mochajs/mocha/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/mochajs/mocha/blob/main/.github/CONTRIBUTING.md) were taken

## Overview

This one fixes a pretty subtle edge case with `timeout(0)`.

when you disable timeouts with `this.timeout(0)`, async tests can end up in a weird situation where they never actually finish but also don’t get stopped by a timeout. If the test forgets to resolve (like a promise never settles or `done()` is never called), the Node event loop just empties out and the process exits with code 0. So it looks like success, even though the test never completed.

to fix that, a `process.on('beforeExit')` listener was added inside `Runnable#run()`. It catches the moment Node is about to shut down while a test is still active, and fails the test with a clear error instead of silently exiting.

Once the test actually finishes, that listener is removed in `done()`, so it only applies to stuck/incomplete tests.

It’s also guarded so it doesn’t run in the browser since `beforeExit` is Node-only.

Tests cover both the runnable behavior and a real integration case using a fixture with a never-resolving promise to simulate the issue. Everything passes, unit tests, integration tests, and TypeScript checks are all clean.

### CODE REVIEW

File: test/integration/fixtures/timeout-zero-silent-exit.fixture.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+describe('timeout 0 silent exit', function () {
+  this.timeout(0);
+
+  it('should fail when promise never resolves', function () {
+    // Return a promise that never resolves.
+    // Without the fix, the process would exit silently with code 0.
```
Comment (JoshuaKGoldberg) [inline test/integration/fixtures/timeout-zero-silent-exit.fixture.js]: [Docs] This is the kind of comment AIs often leave: it thinks in the context of the PR / requested change, not the repo as a whole.

* PR: it makes sense to refer to "the fix", that's what's happening here
* As a whole: what is "the fix"? The meaning of that gets lost over time

As a rule, I'd recommend avoiding these kind of transient comments when possible. A better way to convey them is in good naming, such as the added function in `timeout.spec.js`. That name is really good & informative!

```suggestion
```

File: lib/runnable.js [hunk 2]
```diff
@@ -360,6 +393,11 @@ class Runnable extends EventEmitter {
       var result = fn.call(ctx);
       if (result && typeof result.then === "function") {
         self.resetTimeout();
+        // When timeout is disabled, unref the fallback timer so the event
+        // loop can drain and beforeExit fires for never-resolving promises.
+        if (self.timeout() === 0 && self.timer) {
+          self.timer.unref();
```
Comment (JoshuaKGoldberg) [inline lib/runnable.js]: [Style] Nit: it's just a little cleaner this way ✨ 

```suggestion
        if (self.timeout() === 0) {
          self.timer?.unref();
```

File: lib/runnable.js [hunk 3]
```diff
@@ -316,6 +323,32 @@ class Runnable extends EventEmitter {
...
+            new Error(
+              "async test exited without done() being called or Promise resolving " +
+                "(the event loop has no remaining work); " +
+                "if this test intentionally takes a long time, use a non-zero timeout",
+            ),
+          );
+        }
+      };
+      process.on("beforeExit", beforeExitHandler);
+    }
```
Comment (JoshuaKGoldberg) [inline lib/runnable.js]: [Bug]-ish: this logic is done for all non-`done`-callback tests: both Promise-returning and synchronous. Synchronous tests don't need it. They'll just fire `done()` synchronously.

How about moving this into `callFn()` and checking if the `result` is a thenable?

File: test/integration/timeout.spec.js
Comment (JoshuaKGoldberg) [inline test/integration/timeout.spec.js:1]: [Testing] The one added test here is for a test that returns a Promise. But the the original bug report in #2537 was actually for the [`function(done) {}`]() style. We'll need testing for that too. Which I suspect might reveal some gaps in logic.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
A start 🚀 I think this gets 1 of the 2 needed styles of non-synchronous tests.
